cli/flox-rust-sdk/src/providers/catalog.rs

```diff
@@ -0,0 +1,174 @@
+use std::env;
+
+use catalog_api_v1::types::{self as api_types, error as api_error};
+use catalog_api_v1::{Client as APIClient, Error as APIError};
+use once_cell::sync::Lazy;
+use thiserror::Error;
+
+use crate::data::System;
+
+const DEFAULT_CATALOG_URL: &str = "https://flox-catalog.flox.dev";
+/// Whether to use an actual catalog client or the mock client.
+///
+/// Don't use a feature flag since this shouldn't be exposed to users.
+static USE_CATALOG_MOCK: Lazy<bool> = Lazy::new(|| env::var("_FLOX_USE_CATALOG_MOCK").is_ok());
+
+/// Either a client for the actual catalog service,
+/// or a mock client for testing.
+#[derive(Debug)]
+pub enum Client {
+    Catalog(CatalogClient),
+    Mock(MockClient),
+}
+
+impl Client {
+    pub fn new(mock: bool) -> Self {
+        if mock {
+            Client::Mock(MockClient)
+        } else {
+            Client::Catalog(CatalogClient::default())
+        }
+    }
+}
+
+impl Default for Client {
+    fn default() -> Self {
+        Self::new(*USE_CATALOG_MOCK)
+    }
+}
+
+/// A client for the catalog service.
+///
+/// This is a wrapper around the auto-generated APIClient.
+#[derive(Debug)]
+pub struct CatalogClient {
+    client: APIClient,
+}
+impl CatalogClient {
+    pub fn new() -> Self {
+        Self {
+            client: APIClient::new(DEFAULT_CATALOG_URL),
+        }
+    }
+}
+impl Default for CatalogClient {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+pub struct MockClient;
+
+impl Client {
+    /// Resolve a list of [PackageGroup]s into a list of
+    /// [ResolvedPackageGroup]s.
+    pub async fn resolve(
```

Comment:
exposing the async interfaces seems to become pretty messy. basically everything that does a resolution will need to become async or we got ourselves an async sandwich..